Sources/SourceControl/SwiftPMConfig.swift

```diff
@@ -0,0 +1,123 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2018 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+import Foundation
+
+import Basic
+import Utility
+
+/// Manages a package's configuration.
+///
+// FIXME: We may want to move this class to some other layer once we start
+// supporting more things than just mirrors.
+public final class SwiftPMConfig {
+
+    enum Error: Swift.Error, CustomStringConvertible {
+        case mirrorNotFound
+
+        var description: String {
+            return "mirror not found"
+        }
+    }
+
+    /// Persistence support.
+    let persistence: SimplePersistence?
+
+    /// The schema version of the config file.
+    ///
+    /// * 1: Initial version.
+    static let schemaVersion: Int = 1
+
+    /// The mirrors.
+    private var mirrors: [String: Mirror]
+
+    public init(path: AbsolutePath, fs: FileSystem = localFileSystem) {
+        self.mirrors = [:]
+        self.persistence = SimplePersistence(
+            fileSystem: fs, schemaVersion: SwiftPMConfig.schemaVersion,
+            statePath: path, prettyPrint: true
+        )
+    }
+
+    public init() {
+        self.mirrors = [:]
+        self.persistence = nil
+    }
+
+    /// Set a mirror URL for the given package URL.
+    public func set(mirrorURL: String, forPackageURL packageURL: String) throws {
+        mirrors[packageURL] = Mirror(original: packageURL, mirror: mirrorURL)
+        try saveState()
+    }
+
+    /// Unset a mirror for the given package or mirror URL.
+    ///
+    /// This method will throw if there is no mirror for the given input.
+    public func unset(packageOrMirrorURL: String) throws {
+        if mirrors.keys.contains(packageOrMirrorURL) {
+            mirrors[packageOrMirrorURL] = nil
+        } else if let mirror = mirrors.first(where: { $0.value.mirror == packageOrMirrorURL }) {
+            mirrors[mirror.key] = nil
+        } else {
+            throw Error.mirrorNotFound
+        }
+        try saveState()
+    }
+
+    /// Returns the mirror for the given specificer.
+    public func getMirror(forURL url: String) -> String? {
+        return mirrors[url]?.mirror
+    }
+
+    /// Returns the tr
```

Comment:
Im not sure what the `tr` stands for.

Reply:
Ah, this is a typo. I think I wanted to write something like:

```
/// Returns the mirrored url if it exists, otherwise the original url. 
```

Reply:
Fixed here: https://github.com/apple/swift-package-manager/commit/00e8771d8ddb5f82aaca9efb620dc5d56b4b8cf1